middleware/grpc_cluster.go

```diff
@@ -6,48 +6,139 @@ import (
 
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
-	"github.com/gogo/status"
+
+	"github.com/grafana/dskit/clusterutil"
+	"github.com/grafana/dskit/grpcutil"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/metadata"
-)
-
-const (
-	MetadataClusterKey = "x-cluster"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
 )
 
-// ClusterUnaryClientInterceptor propagates the given cluster info to gRPC metadata.
-func ClusterUnaryClientInterceptor(cluster string) grpc.UnaryClientInterceptor {
+// ClusterUnaryClientInterceptor propagates the given cluster label to gRPC metadata, before calling the next invoker.
+// If an empty cluster label, nil invalidCounter or nil logger are provided, ClusterUnaryClientInterceptor panics.
+// If the softValidation parameter is true, errors related to the cluster label validation are logged, but not returned.
+// Otherwise, these errors are returned, and invalidCounter is incremented.
+func ClusterUnaryClientInterceptor(cluster string, softValidation bool, invalidCluster *prometheus.CounterVec, logger log.Logger) grpc.UnaryClientInterceptor {
+	validateClusterClientInterceptorInputParameters(cluster, invalidCluster, logger)
 	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
-		if cluster != "" {
-			ctx = metadata.AppendToOutgoingContext(ctx, MetadataClusterKey, cluster)
+		// We skip the gRPC health check.
+		if method == healthpb.Health_Check_FullMethodName {
+			return invoker(ctx, method, req, reply, cc, opts...)
 		}
 
-		return invoker(ctx, method, req, reply, cc, opts...)
+		msgs, err := checkClusterFromIncomingContext(ctx, method, cluster, true, softValidation)
+		if len(msgs) > 0 {
+			level.Warn(logger).Log(msgs...)
+		}
+		if err != nil {
+			invalidCluster.WithLabelValues(method, clusterutil.ReasonClient).Inc()
+			return grpcutil.Status(codes.Internal, err.Error()).Err()
+		}
```

Comment:
I'm confused about this, this looks like the client is checking the `cluster` _before_ sending the request? Did I understand correctly?